### PR TITLE
Add parallel linter

### PR DESCRIPTION
**Pull request by roysagi**
Added run_parallel_pkg_dev_tasks.py functionality to lint in demisto-sdk.
Using LintManager and -p flag to run several packages in parallel.

### CODE REVIEW

File: demisto_sdk/core.py [hunk 1]
```diff
@@ -42,6 +43,7 @@ def initialize_parsers(self):
         Extractor.add_sub_parser(self.subparsers)
         FilesValidator.add_sub_parser(self.subparsers)
         Linter.add_sub_parser(self.subparsers)
```
Comment (liorblob) [inline demisto_sdk/core.py]: Not sure about the renaming, but you forgot to remove this line

File: demisto_sdk/dev_tools/lint_manager.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        self.keep_container = keep_container
+        self.cpu_num = cpu_num
+        self.common_server_created = False
+        self.run_args = {
+            'pylint': not no_pylint,
+            'flake8': not no_flake8,
+            'mypy': not no_mypy,
+            'tests': not no_test
+        }
+        self.pkgs = project_dir.split(',')
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: need to specify that this param will support a list as well, as the name does not indicate it.

File: demisto_sdk/dev_tools/lint_manager.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        self.configuration = configuration
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: what's that

File: demisto_sdk/dev_tools/lint_manager.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    def run_dev_packages(self) -> int:
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: docstriing

File: demisto_sdk/dev_tools/lint_manager.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        if not self.parallel:
+            linter = Linter(self.pkgs[0], no_test=not self.run_args['tests'],
+                            no_pylint=not self.run_args['pylint'], no_flake8=not self.run_args['flake8'],
+                            no_mypy=not self.run_args['mypy'], verbose=self.log_verbose, root=self.root,
+                            keep_container=self.keep_container, cpu_num=self.cpu_num, configuration=self.configuration)
+
+            return linter.run_dev_packages()
+
+        # we run parallel processes
+        else:
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: extract to another function

File: demisto_sdk/dev_tools/lint_manager.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            max_workers = int(os.getenv("MAX_WORKERS", "10"))
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: why is this not a script param?

File: demisto_sdk/dev_tools/lint_manager.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                    print_color("\n******* SUCCESS PKGS: *******", LOG_COLORS.GREEN)
+                    print_color("\n\t{}\n".format("\n\t".join(good_pkgs)), LOG_COLORS.GREEN)
+
+                if not good_pkgs and not fail_pkgs:
+                    print_color("\n******* No changed packages found *******\n", LOG_COLORS.YELLOW)
+
+                if fail_pkgs:
+                    sys.exit(1)
+
+    def _should_run_pkg(self, pkg_dir: str) -> bool:
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: not a good name, not sure what it does, add docstring

File: demisto_sdk/dev_tools/lint_manager.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            return linter.run_dev_packages()
+
+        # we run parallel processes
+        else:
+            max_workers = int(os.getenv("MAX_WORKERS", "10"))
+            pkg_dirs = self.pkgs
+            pkgs_to_run = []
+            for dir in pkg_dirs:
+                if self._should_run_pkg(pkg_dir=dir):
+                    pkgs_to_run.append(dir)
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: seems like a function to me

File: demisto_sdk/dev_tools/lint_manager.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+                if good_pkgs:
+                    print_color("\n******* SUCCESS PKGS: *******", LOG_COLORS.GREEN)
+                    print_color("\n\t{}\n".format("\n\t".join(good_pkgs)), LOG_COLORS.GREEN)
+
+                if not good_pkgs and not fail_pkgs:
+                    print_color("\n******* No changed packages found *******\n", LOG_COLORS.YELLOW)
+
+                if fail_pkgs:
+                    sys.exit(1)
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: function

File: demisto_sdk/dev_tools/lint_manager.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                pkgs_to_run.remove('Scripts/CommonServerPython')
+                res, _ = self._run_single_package_thread(package_dir='Scripts/CommonServerPython')
+                if res == 0:
+                    good_pkgs.append('Scripts/CommonServerPython')
+
+                else:
+                    fail_pkgs.append('Scripts/CommonServerPython')
+
+            with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+                futures_submit = [executor.submit(self._run_single_package_thread, dir) for dir in pkgs_to_run]
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: dir is a python saved word, lets rename this arg

File: demisto_sdk/dev_tools/lint_manager.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    def _run_single_package_thread(self, package_dir) -> Tuple[int, str]:
+        linter = Linter(package_dir, no_test=not self.run_args['tests'],
+                        no_pylint=not self.run_args['pylint'], no_flake8=not self.run_args['flake8'],
+                        no_mypy=not self.run_args['mypy'], verbose=self.log_verbose, root=self.root,
+                        keep_container=self.keep_container, cpu_num=self.cpu_num, configuration=self.configuration)
+
+        return linter.run_dev_packages(), package_dir
+
+    @staticmethod
+    def add_sub_parser(subparsers):
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/lint_manager.py]: where is this being used?

File: demisto_sdk/dev_tools/lint_manager.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                            keep_container=self.keep_container, cpu_num=self.cpu_num, configuration=self.configuration)
+
+            return linter.run_dev_packages()
+
+        # we run parallel processes
+        else:
+            max_workers = int(os.getenv("MAX_WORKERS", "10"))
+            pkg_dirs = self.pkgs
+            pkgs_to_run = []
+            for dir in pkg_dirs:
```
Comment (yaakovi) [inline demisto_sdk/dev_tools/lint_manager.py]: don't shadow builtin functions (`dir`)

File: demisto_sdk/core.py [hunk 13]
```diff
@@ -139,8 +139,8 @@ def lint(self, project_dir: str, **kwargs):
         :param kwargs Optional arguments.
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/core.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 2 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 3 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: I took this from the Linter object, It is passed through the core file to every class initiated.

File: demisto_sdk/dev_tools/lint_manager.py (hunk 4 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 5 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 6 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: added

File: demisto_sdk/dev_tools/lint_manager.py (hunk 12 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 8 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 10 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: changed

File: demisto_sdk/dev_tools/lint_manager.py (hunk 9 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 7 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

File: demisto_sdk/dev_tools/lint_manager.py (hunk 11 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: in the core.py part - this defines all flags that are used with Lint.

File: demisto_sdk/dev_tools/linter.py [hunk 14]
```diff
@@ -97,7 +97,11 @@ def run_dev_packages(self) -> int:
         for docker in dockers:
             for try_num in (1, 2):
                 print_v("Using docker image: {}".format(docker))
+                LOCK.acquire(blocking=True)
+                print("========For: {} ========".format(self.project_dir))
                 py_num = get_python_version(docker, self.log_verbose)
+                print("")
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/linter.py]: done

File: demisto_sdk/dev_tools/linter.py [hunk 15]
```diff
@@ -114,35 +118,48 @@ def run_dev_packages(self) -> int:
                     if not self.log_verbose:
                         sys.stderr.write("Need a more detailed log? try running with the -v options as so: \n{} -v\n"
                                          .format(" ".join(sys.argv[:])))
+
+                    print("")
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/linter.py]: done

File: demisto_sdk/dev_tools/linter.py [hunk 16]
```diff
@@ -114,35 +118,48 @@ def run_dev_packages(self) -> int:
...
+        LOCK.acquire(blocking=True)
+        lint_files = self._get_lint_files()
+        print("========= Running flake8 on: {}===============".format(lint_files))
         python_exe = 'python2' if py_num < 3 else 'python3'
         print_v('Using: {} to run flake8'.format(python_exe))
         sys.stdout.flush()
         subprocess.check_call([python_exe, '-m', 'flake8', self.project_dir], cwd=self.configuration.env_dir)
-        print("flake8 completed")
+        print("flake8 completed for: {}".format(lint_files))
+        print("")
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/linter.py]: done

File: demisto_sdk/dev_tools/linter.py [hunk 17]
```diff
@@ -114,35 +118,48 @@ def run_dev_packages(self) -> int:
...
         try:
             self.get_common_server_python()
             lint_files = self._get_lint_files()
             print("========= Running mypy on: {} ===============".format(lint_files))
             sys.stdout.flush()
             script_path = os.path.abspath(os.path.join(self.configuration.sdk_env_dir, self.run_mypy_script))
             subprocess.check_call(['bash', script_path, str(py_num), lint_files], cwd=self.project_dir)
-            print("mypy completed")
+            print("mypy completed for: {}".format(lint_files))
+            print("")
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/linter.py]: done

File: demisto_sdk/dev_tools/thread_lock.py [hunk 18]
```diff
@@ -0,0 +1,3 @@
+import threading
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/thread_lock.py]: deleted

File: demisto_sdk/dev_tools/lint_manager.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+import subprocess
+import concurrent.futures
+from typing import Tuple
+from demisto_sdk.common.configuration import Configuration
+from demisto_sdk.common.tools import print_color, LOG_COLORS
+
+from demisto_sdk.dev_tools.linter import Linter
+
+
+class LintManager:
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/lint_manager.py]: done

**Comment by roysagi** [comment]
We need a lock to make the results easily readable.

Without it, all the prints are meshed together and it is tough to understand.

I tried masking the stdout and stderr but it didn't work, so only locking sorted this out.

**Pull request closed without merging**